From ff970d041a87719c32e50e1f17f93d00a6f3ec58 Mon Sep 17 00:00:00 2001
From: Till Hofmann <hofmann@kbsg.rwth-aachen.de>
Date: Fri, 30 Apr 2021 16:00:01 +0200
Subject: [PATCH] Install files into /usr/libexec/bear instead of /usr/libexec
 (#348)

* Install files into /usr/libexec/bear instead of directly into /usr/libexec

As the installed files have rather generic names (e.g., libexec.so), put them
into a sub-directory so it is clear that those files belong to bear.

* Use PRIVATE_INSTALLDIR for private binaries

This defaults to `${LIBEXECDIR}/bear`, i.e., `/usr/libexec/bear`.

Use the directory for all binaries that should not be in PATH or in the
default link directory.
---
 CMakeLists.txt                  | 4 ++++
 source/config.h.in              | 6 +++---
 source/intercept/CMakeLists.txt | 8 ++++----
 test/CMakeLists.txt             | 6 +++---
 4 files changed, 14 insertions(+), 10 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2a71eca..1820787 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -33,6 +33,8 @@ if (NOT CMAKE_INSTALL_LIBEXECDIR)
     include(GNUInstallDirs)
 endif ()
 
+set(PRIVATE_INSTALLDIR "${CMAKE_INSTALL_LIBEXECDIR}/bear" CACHE PATH "Install directory for private binaries")
+
 # Verify or install dependencies
 add_subdirectory(third_party)
 
@@ -58,6 +60,7 @@ ExternalProject_Add(BearSource
             -DCMAKE_INSTALL_PREFIX:PATH=${STAGED_INSTALL_PREFIX}
             -DCMAKE_INSTALL_LIBEXECDIR:PATH=${CMAKE_INSTALL_LIBEXECDIR}
             -DROOT_INSTALL_PREFIX:PATH=${CMAKE_INSTALL_PREFIX}
+            -DPRIVATE_INSTALLDIR:PATH=${PRIVATE_INSTALLDIR}
             -DCMAKE_BUILD_TYPE:STRING=${CMAKE_BUILD_TYPE}
             -DENABLE_UNIT_TESTS:BOOL=${ENABLE_UNIT_TESTS}
             -DENABLE_FUNC_TESTS:BOOL=${ENABLE_FUNC_TESTS}
@@ -77,6 +80,7 @@ if (ENABLE_FUNC_TESTS)
             CMAKE_CACHE_ARGS
                 -DSTAGED_INSTALL_PREFIX:PATH=${STAGED_INSTALL_PREFIX}
                 -DBEARSOURCE_BUILD_DIR:PATH=${CMAKE_BINARY_DIR}/subprojects/Build/BearSource
+                -DPRIVATE_INSTALLDIR:PATH=${PRIVATE_INSTALLDIR}
             TEST_BEFORE_INSTALL
                 1
             INSTALL_COMMAND
diff --git a/source/config.h.in b/source/config.h.in
index 348ec92..ffcce3a 100644
--- a/source/config.h.in
+++ b/source/config.h.in
@@ -107,8 +107,8 @@ namespace cmd {
     }
 
     namespace wrapper {
-        constexpr char DEFAULT_PATH[] = "@ROOT_INSTALL_PREFIX@/@CMAKE_INSTALL_LIBEXECDIR@/wrapper";
-        constexpr char DEFAULT_DIR_PATH[] = "@ROOT_INSTALL_PREFIX@/@CMAKE_INSTALL_LIBEXECDIR@/wrapper.d";
+        constexpr char DEFAULT_PATH[] = "@ROOT_INSTALL_PREFIX@/@PRIVATE_INSTALLDIR@/wrapper";
+        constexpr char DEFAULT_DIR_PATH[] = "@ROOT_INSTALL_PREFIX@/@PRIVATE_INSTALLDIR@/wrapper.d";
 
         constexpr char FLAG_VERBOSE[] = "--verbose";
         constexpr char FLAG_DESTINATION[] = "--destination";
@@ -120,7 +120,7 @@ namespace cmd {
     }
 
     namespace library {
-        constexpr char DEFAULT_PATH[] = "@ROOT_INSTALL_PREFIX@/@CMAKE_INSTALL_LIBEXECDIR@/@CMAKE_SHARED_LIBRARY_PREFIX@exec@CMAKE_SHARED_LIBRARY_SUFFIX@";
+        constexpr char DEFAULT_PATH[] = "@ROOT_INSTALL_PREFIX@/@PRIVATE_INSTALLDIR@/@CMAKE_SHARED_LIBRARY_PREFIX@exec@CMAKE_SHARED_LIBRARY_SUFFIX@";
 
         constexpr char KEY_REPORTER[] = "INTERCEPT_REPORT_COMMAND";
         constexpr char KEY_DESTINATION[] = "INTERCEPT_REPORT_DESTINATION";
diff --git a/source/intercept/CMakeLists.txt b/source/intercept/CMakeLists.txt
index 142080e..73fef86 100644
--- a/source/intercept/CMakeLists.txt
+++ b/source/intercept/CMakeLists.txt
@@ -106,9 +106,9 @@ target_link_libraries(wrapper
         wrapper_a)
 
 install(TARGETS wrapper
-        RUNTIME DESTINATION ${CMAKE_INSTALL_LIBEXECDIR})
+  RUNTIME DESTINATION ${PRIVATE_INSTALLDIR})
 install(DIRECTORY
-        DESTINATION ${CMAKE_INSTALL_LIBEXECDIR}/wrapper.d)
+  DESTINATION ${PRIVATE_INSTALLDIR}/wrapper.d)
 install(CODE "
     execute_process(
         COMMAND ln -sf ../wrapper ar
@@ -142,7 +142,7 @@ install(CODE "
         COMMAND ln -sf ../wrapper objdump
         COMMAND ln -sf ../wrapper ranlib
         COMMAND ln -sf ../wrapper readelf
-        WORKING_DIRECTORY ${DESTDIR}${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBEXECDIR}/wrapper.d
+        WORKING_DIRECTORY ${DESTDIR}${CMAKE_INSTALL_PREFIX}/${PRIVATE_INSTALLDIR}/wrapper.d
     )
 ")
 
@@ -187,7 +187,7 @@ elseif (CMAKE_SYSTEM_NAME STREQUAL "Darwin")
 endif ()
 
 install(TARGETS exec
-        LIBRARY DESTINATION ${CMAKE_INSTALL_LIBEXECDIR})
+  LIBRARY DESTINATION ${PRIVATE_INSTALLDIR})
 
 
 # Create unit test.
diff --git a/test/CMakeLists.txt b/test/CMakeLists.txt
index e2d5d52..d73a106 100644
--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -53,9 +53,9 @@ if (LIT_EXECUTABLE)
             -D_BIN_BEAR=${STAGED_INSTALL_PREFIX}/${CMAKE_INSTALL_BINDIR}/bear
             -D_BIN_CITNAMES=${STAGED_INSTALL_PREFIX}/${CMAKE_INSTALL_BINDIR}/citnames
             -D_BIN_INTERCEPT=${STAGED_INSTALL_PREFIX}/${CMAKE_INSTALL_BINDIR}/intercept
-            -D_BIN_ER=${STAGED_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBEXECDIR}/wrapper
-            -D_BIN_WRAPPER=${STAGED_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBEXECDIR}/wrapper.d
-            -D_LIB_EXEC=${STAGED_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBEXECDIR}/${CMAKE_SHARED_LIBRARY_PREFIX}exec${CMAKE_SHARED_LIBRARY_SUFFIX}
+            -D_BIN_ER=${STAGED_INSTALL_PREFIX}/${PRIVATE_INSTALLDIR}/wrapper
+            -D_BIN_WRAPPER=${STAGED_INSTALL_PREFIX}/${PRIVATE_INSTALLDIR}/wrapper.d
+            -D_LIB_EXEC=${STAGED_INSTALL_PREFIX}/${PRIVATE_INSTALLDIR}/${CMAKE_SHARED_LIBRARY_PREFIX}exec${CMAKE_SHARED_LIBRARY_SUFFIX}
             -D_BIN_EVENTS_DB=${BEARSOURCE_BUILD_DIR}/intercept/events_db
             -v
             ${CMAKE_CURRENT_BINARY_DIR})
-- 
2.31.1

